### PR TITLE
EXP-660: Payload Experiment MVP

### PR DESCRIPTION
**Pull request by zacharygoodwin**

### CODE REVIEW

File: proctor-common/src/main/java/com/indeed/proctor/common/model/PayloadExperimentConfig.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Collections;
+import java.util.List;
+
+@Data
+@Builder
+public class PayloadExperimentConfig {
+    @Nullable private String priority;
```
Comment (vanhaxl) [inline proctor-common/src/main/java/com/indeed/proctor/common/model/PayloadExperimentConfig.java:14]: Is `priority` a String? Should it be an `integer`?

Comment (zacharygoodwin) [reply proctor-common/src/main/java/com/indeed/proctor/common/model/PayloadExperimentConfig.java:14]: Should be a string. The matrix stores the version as a string which will be used to populate. There was desire for also custom input and lexico sort for priority

Comment (jamesallen-indeed) [reply proctor-common/src/main/java/com/indeed/proctor/common/model/PayloadExperimentConfig.java:20]: Just to confirm, we want lexico and **not** ordinal compare for `AppVersion` (for priority)?

**Review by jamesallen-indeed — APPROVED** [review]
LGTM; just a simple question

File: proctor-common/src/main/java/com/indeed/proctor/common/model/PayloadExperimentConfig.java (hunk 1 above)
Comment (zacharygoodwin) [reply proctor-common/src/main/java/com/indeed/proctor/common/model/PayloadExperimentConfig.java:20]: Hmmm, good catch right now should be ordinal but for the MVP phase 2 when we allow custom string we will want to allow lexico.

File: proctor-common/src/main/java/com/indeed/proctor/common/Proctor.java [hunk 2]
```diff
@@ -515,4 +523,56 @@ private boolean isIncognitoEnabled(@Nonnull final Map<String, Object> inputConte
...
+            final Map<String, PayloadProperty> testProperties) {
+        final PayloadProperty curr = testProperties.get(field.getKey());
+        // store property if it does not exist in map
+        if (curr == null) {
+            testProperties.put(
+                    field.getKey(),
+                    PayloadProperty.builder().value(field.getValue()).testName(testName).build());
+        } else {
+            final PayloadExperimentConfig currPayloadConfig =
+                    testChoosers
```
Comment (stevenchen-indeed) [inline proctor-common/src/main/java/com/indeed/proctor/common/Proctor.java:557]: nit: should we extract common code to 
getPayloadExperimentConfigFromTestDefinition(testName)?

File: proctor-common/src/main/java/com/indeed/proctor/common/Proctor.java [hunk 3]
```diff
@@ -515,4 +523,56 @@ private boolean isIncognitoEnabled(@Nonnull final Map<String, Object> inputConte
...
+                            .get(curr.getTestName())
+                            .getTestDefinition()
+                            .getPayloadExperimentConfig();
+            final PayloadExperimentConfig newPayloadConfig =
+                    testChoosers.get(testName).getTestDefinition().getPayloadExperimentConfig();
+            // store property if it has higher priority than the currently stored property
+            try {
+                if (isHigherPriority(currPayloadConfig, newPayloadConfig)) {
```
Comment (stevenchen-indeed) [inline proctor-common/src/main/java/com/indeed/proctor/common/Proctor.java:565]: nit: 

it is great that we extracted a method to compare priority.
However it is not clear from reading this line when it returns true, it is the first or second parameter being higher priority(i recognize that my statement can be debatable). 

Will it be better to extract a method call getWinningPropertyValue which will get the highest priority value, which can be existing value or the new value. if the existing value does not exist, then the new value is the winning value. I think it is ok to reinsert the existing value back(to make code shorter) if it is winning again. 

Do we expect priority checking to be evolving? If yes, should we have an interface for priority checker and this class only cares about invoking the interface?

File: proctor-common/src/main/java/com/indeed/proctor/common/model/PayloadExperimentConfig.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        return "PayloadExperimentConfig{"
+                + "priority='"
+                + priority
+                + '\''
+                + ", namespaces="
+                + namespaces
+                + '}';
+    }
+
+    public static boolean isHigherPriority(
```
Comment (stevenchen-indeed) [inline proctor-common/src/main/java/com/indeed/proctor/common/model/PayloadExperimentConfig.java:34]: if this is instance method, we can do thisPayloadExperimentConfig.isHigherPriorityThan(otherPayloadConfig).
Any reason we want it to be static?

File: proctor-common/src/test/java/com/indeed/proctor/common/dynamic/TestNamespacesFilter.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+import com.google.common.collect.ImmutableSet;
+import com.indeed.proctor.common.Serializers;
+import com.indeed.proctor.common.model.ConsumableTestDefinition;
+import com.indeed.proctor.common.model.PayloadExperimentConfig;
+import com.indeed.proctor.common.model.TestType;
+import org.junit.Test;
+
+import java.util.List;
+
+import static java.util.Collections.*;
```
Comment (stevenchen-indeed) [inline proctor-common/src/test/java/com/indeed/proctor/common/dynamic/TestNamespacesFilter.java:15]: nit: use single class imports.

File: proctor-common/src/main/java/com/indeed/proctor/common/Proctor.java (hunk 3 above)
Comment (zacharygoodwin) [reply proctor-common/src/main/java/com/indeed/proctor/common/Proctor.java:565]: I think we can address the interface when we go to do/determine the exact need for the more advanced comparison.

Comment (zacharygoodwin) [reply proctor-common/src/main/java/com/indeed/proctor/common/Proctor.java:565]: Will do the shorten code suggestion

File: proctor-common/src/test/java/com/indeed/proctor/common/dynamic/TestNamespacesFilter.java (hunk 5 above)
Comment (zacharygoodwin) [reply proctor-common/src/test/java/com/indeed/proctor/common/dynamic/TestNamespacesFilter.java:15]: Fixed

File: proctor-common/src/main/java/com/indeed/proctor/common/model/PayloadExperimentConfig.java (hunk 4 above)
Comment (zacharygoodwin) [reply proctor-common/src/main/java/com/indeed/proctor/common/model/PayloadExperimentConfig.java:34]: Moved to class method

File: proctor-common/src/main/java/com/indeed/proctor/common/Proctor.java (hunk 3 above)
Comment (zacharygoodwin) [reply proctor-common/src/main/java/com/indeed/proctor/common/Proctor.java:565]: I choose to make it a class/instance method so will keep the code as is in respect to shortening the code as no longer applicable.

File: proctor-common/src/main/java/com/indeed/proctor/common/Proctor.java (hunk 2 above)
Comment (zacharygoodwin) [reply proctor-common/src/main/java/com/indeed/proctor/common/Proctor.java:557]: I personally do not think it is worth a method. Not that much duplication

**Pull request merged**